cpp/src/copying/gather.cu

```diff
@@ -360,38 +266,104 @@ void gather(table const* source_table, gdf_index_type const gather_map[],
   CUDF_EXPECTS(nullptr != gather_map, "gather_map is null");
   CUDF_EXPECTS(source_table->num_columns() == destination_table->num_columns(),
                "Mismatched number of columns");
+  const gdf_size_type n_cols = source_table->num_columns();
+
+  // We create `n_cols` streams for the `n_cols` kernels we are gonna launch.
+  std::vector<util::cuda::scoped_stream> v_stream(n_cols);
 
-  auto gather_column = [gather_map, check_bounds, stream](
-                           gdf_column const* source, gdf_column* destination) {
-    CUDF_EXPECTS(source->dtype == destination->dtype, "Column type mismatch");
+  for (gdf_size_type i = 0; i < n_cols; i++) {
+    // Perform sanity checks
+    gdf_column *dest_col = destination_table->get_column(i);
+    const gdf_column *src_col = source_table->get_column(i);
 
-    // If the source column has a valid buffer, the destination column must
-    // also have one
-    bool const source_has_nulls{source->valid != nullptr};
-    bool const dest_has_nulls{destination->valid != nullptr};
-    CUDF_EXPECTS((source_has_nulls && dest_has_nulls) || (not source_has_nulls),
-                 "Missing destination validity buffer");
+    CUDF_EXPECTS(src_col->dtype == dest_col->dtype, "Column type mismatch");
 
-    // TODO: Each column could be gathered on a separate stream
-    cudf::type_dispatcher(source->dtype, column_gatherer{}, source, gather_map,
-                          destination, check_bounds, stream);
+    CUDF_EXPECTS(dest_col->data != nullptr, "Missing source data buffer.");
+
+    // The data gather for n columns will be put on the first n streams
+    cudf::type_dispatcher(src_col->dtype, column_gatherer{}, src_col,
+                          gather_map, dest_col, check_bounds, v_stream[i],
+                          merge_nvstring_category);
+
+    if (cudf::is_nullable(*src_col)) {
+      CUDF_EXPECTS(cudf::is_nullable(*dest_col),
+                   "Missing destination null mask.");
+    }
+  }
 
-    return destination;
-  };
+  rmm::device_vector<gdf_size_type> d_count_vec(n_cols, 0);
```

Comment:
Let's use a more descriptive name. No need for the `d_` prefix here since it's a `device_vector` makes it clear that it is device memory. 
```suggestion
  rmm::device_vector<gdf_size_type> null_counts(n_cols, 0);
```